### PR TITLE
feat(behavior_path_planner): add new manager independent from behavior tree

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

:arrow_down: This must merge before this PR.
https://github.com/autowarefoundation/autoware_launch/pull/218

In this PR, I add new manager that doesn't depend on behavior tree library. The manager is implemented based on the new architecture on [this](https://github.com/orgs/autowarefoundation/discussions/3097) discussion.

### NOTE

- Conventional manager that based on BT library still remains, and that manager is activated by default.
- Change manager in CMakelist.
- The new manager is not compatible with the old one. We have to add the new scene modules that support new manager.
- In this PR, there is no scene modules that support new manager. The new manager can only execute lane following.
- The new scene modules will be added in the other PRs.

## Related links

- [discussion](https://github.com/orgs/autowarefoundation/discussions/3097)
- [autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/218)

### Evaluator

[[TIER IV INTERNAL LINK]](https://evaluation.tier4.jp/evaluation/reports/5ab36667-0cdb-51db-a50b-186b2b59f175?project_id=prd_jt) Baseline 811/814
[[TIER IV INTERNAL LINK]](https://evaluation.tier4.jp/evaluation/reports/b8cfb8fc-40ef-503b-88c4-aa00780bc722?project_id=prd_jt) This PR 812/814

### Test by Psim

Please cherry-pick [this](https://github.com/autowarefoundation/autoware_launch/pull/218) commit to your `autoware_launch` or add following line to `behavior_path_planner.param.yaml`.

```yaml
/**:
  ros__parameters:
    verbose: true # this line
   ...
```

If you want to use new manager, please set **USE_BT** to `FALSE`. For now, the manager can execute only lane following.

```cmake
# use planner manager that supports behavior tree
set(USE_BT FALSE)
message(STATUS "USE_BEHAVIOR_TREE: " ${USE_BT})
```

https://user-images.githubusercontent.com/44889564/220276091-0b1fa05a-5908-4e15-9f8b-4cbe85d32ecc.mp4

If you want to use conventional manager, please set **USE_BT** to `TRUE`.

```cmake
# use planner manager that supports behavior tree
set(USE_BT TRUE)
message(STATUS "USE_BEHAVIOR_TREE: " ${USE_BT})
```

https://user-images.githubusercontent.com/44889564/220276416-106ee601-674c-4873-aba9-d73ae99502bd.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  std::vector<std::shared_ptr<SceneModuleManagerInterface>> getSceneModuleManagers() const
+  {
+    return scene_manager_ptrs_;
+  }
+
+  std::vector<std::shared_ptr<SceneModuleStatus>> getSceneModuleStatus() const
+  {
+    std::vector<std::shared_ptr<SceneModuleStatus>> ret;
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp:101]: It seems we can reserve this vector in advance.

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+  virtual void removeRTCStatus()
+  {
+    if (!rtc_interface_ptr_) {
+      return;
+    }
+    rtc_interface_ptr_->clearCooperateStatus();
+  }
+
+  void alreadyApproved() { is_already_approved_ = true; }
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp:268]: Function name should begin from verb.
So, I'd like to change the name appropriately, but I don't understand the nuance of "already".
Could you tell me the reason of this naming?

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  void nothingApproved() { is_already_approved_ = false; }
+
+  void waitApproval() { is_waiting_approval_ = true; }
+
+  void clearWaitingApproval() { is_waiting_approval_ = false; }
+
+  rclcpp::Clock::SharedPtr clock_;
+
+  std::shared_ptr<RTCInterface> rtc_interface_ptr_;
+  std::unique_ptr<SteeringFactorInterface> steering_factor_interface_ptr_;
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp:279]: `rtc_interface` and `steering_factor_interface` have a publisher, so `SceneModuleManager` should have these interfaces, not `SceneModuleInterface`.

The implementation in `behavior_velocity_planner` may be helpful.
https://github.com/autowarefoundation/autoware.universe/blob/b28d792dee2379b7e9561ec256dc0052c60d127b/planning/behavior_velocity_planner/include/scene_module/scene_module_interface.hpp#L345-L430

(There is no problem if only one scene module is registered. If it is hard to change, it is okay to do it later.)

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    return ret;
+  }
+
+  void resetRootLanelet(const std::shared_ptr<PlannerData> & data);
+
+  void print() const;
+
+private:
+  BehaviorModuleOutput run(
+    const ModuleID & id, const BehaviorModuleOutput & previous_module_output) const
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp:130]: Does using `ModuleID` as one of the arguments mean that planner manager manages UUID?
If it is true, I think planner manager should manage one of the _status_ from scene modules, not UUID.

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp (hunk 2 above)
Comment (rej55) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp:268]: for example, `guard`, `protect`, `block`, `interrupt` ...

Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp:268]:
```c++
bool is_blocked_other_module_launch_;

bool islockedOtherModuleLaunch() { return is_blocked_new_module_launch_; }
void lockedOtherModuleLaunch() { return is_blocked_new_module_launch_ = true; }
void unlockedOtherModuleLaunch() { return is_blocked_new_module_launch_ = false; }
```

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp:94]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2919/commits/e907f5c0e12d36a96cfaa72df24327fd41c74e48.

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp (hunk 2 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_interface.hpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2919/commits/88d631ec4f903a21553528c3f75674ca2bfdc612.

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp (hunk 4 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2919/commits/00be57682e73ac2a98333bbba6f56ab681523ea8.

**Review by rej55 — APPROVED** [review]
LGTM

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,452 @@
+// Copyright 2022 TIER IV, Inc.
```
Comment (zulfaqar-azmi-t4) [inline planning/behavior_path_planner/src/planner_manager.cpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_manager_interface.hpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
+
+// Copyright 2022 TIER IV, Inc.
```
Comment (zulfaqar-azmi-t4) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module_v2/scene_module_manager_interface.hpp:2]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 TIER IV, Inc.
```
Comment (zulfaqar-azmi-t4) [inline planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

**Comment by satoshi-ota** [comment]
@rej55 @zulfaqar-azmi-t4 There is no critical regression.

[[TIER IV INTERNAL LINK]](https://evaluation.tier4.jp/evaluation/reports/5ab36667-0cdb-51db-a50b-186b2b59f175?project_id=prd_jt) Baseline 811/814
[[TIER IV INTERNAL LINK]](https://evaluation.tier4.jp/evaluation/reports/b8cfb8fc-40ef-503b-88c4-aa00780bc722?project_id=prd_jt) This PR 812/814

**Pull request merged**